ui/src/features/Editor/useCanvasCopyPaste.ts

```diff
@@ -221,46 +249,176 @@ export default ({
       edges: edges.filter((e) => e.selected),
     };
     let referencedWorkflows: Workflow[] = [];
-    if (selected.nodes.some((n) => n.type === "reader"))
+
+    if (selected.nodes.length === 0 && selected.edges.length === 0) return;
+
+    const { nodes: nodesToCopy, edges: edgesToCopy } = processNodesAndEdges(
+      selected.nodes,
+      selected.edges,
+    );
+
+    if (nodesToCopy.some((n) => n.type === "subworkflow")) {
+      referencedWorkflows = collectSubworkflows(nodesToCopy, rawWorkflows);
+      if (referencedWorkflows.length === 0) return;
+    }
+
+    if (nodesToCopy.some((n) => n.type === "reader")) {
       return toast({
         title: t("Reader node cannot be copied"),
         description: t("Only one reader can be present in any project."),
         variant: "default",
       });
-
-    if (selected.nodes.length === 0 && selected.edges.length === 0) return;
-
-    if (selected.nodes.some((n) => n.type === "subworkflow")) {
-      referencedWorkflows = collectSubworkflows(selected.nodes, rawWorkflows);
-      if (referencedWorkflows.length === 0) return;
     }
 
     await copy({
-      nodes: selected.nodes,
-      edges: selected.edges,
+      nodes: nodesToCopy,
+      edges: edgesToCopy,
       workflows: referencedWorkflows,
       copiedAt: Date.now(),
     });
-  }, [nodes, edges, collectSubworkflows, copy, rawWorkflows, toast, t]);
+  }, [
+    nodes,
+    edges,
+    collectSubworkflows,
+    copy,
+    processNodesAndEdges,
+    rawWorkflows,
+    toast,
+    t,
+  ]);
+
+  const handleCut = useCallback(
+    async (isCutByShortCut?: boolean) => {
+      const selected: { nodes: Node[]; edges: Edge[] } | undefined = {
+        nodes: nodes.filter((n) => n.selected),
+        edges: edges.filter((e) => e.selected),
+      };
+      if (selected.nodes.length === 0 && selected.edges.length === 0) return;
 
+      const { nodes: nodesToCut, edges: edgesToCut } = processNodesAndEdges(
+        selected.nodes,
+        selected.edges,
+      );
+
+      let referencedWorkflows: Workflow[] = [];
+      if (nodesToCut.some((n) => n.type === "subworkflow")) {
+        referencedWorkflows = collectSubworkflows(nodesToCut, rawWorkflows);
+        if (referencedWorkflows.length === 0) return;
+      }
+
+      await copy({
+        nodes: nodesToCut,
+        edges: edgesToCut,
+        workflows: referencedWorkflows,
+        copiedAt: Date.now(),
+        isCutByShortCut,
+      });
+
+      const nodeChanges: NodeChange[] = nodesToCut.map((n) => ({
+        id: n.id,
+        type: "remove",
+      }));
+
+      const edgeChanges: EdgeChange[] = edgesToCut.map((e) => ({
+        id: e.id,
+        type: "remove",
+      }));
+
+      handleNodesChange(nodeChanges);
+      handleEdgesChange(edgeChanges);
+    },
+    [
+      nodes,
+      edges,
+      processNodesAndEdges,
+      handleNodesChange,
+      handleEdgesChange,
+      collectSubworkflows,
+      copy,
+      rawWorkflows,
+    ],
+  );
```

Comment:
This is 95% what handleCopy is. What you can do is return the nodes and edges in handleCopy, and use handleCopy in handleCut to reduce a lot of code